File: datasets/python_source-code/Multiple-C2/organic_3.py
"""
0/1 Knapsack (DP) - Organic Submission #3
Top-down recursive solution with a manual memoization dictionary.
"""


def solve(index, remaining_capacity, weights, values, memo):
    if index == len(weights) or remaining_capacity == 0:
        return 0

    key = (index, remaining_capacity)
    if key in memo:
        return memo[key]

    # option 1: skip current item
    best = solve(index + 1, remaining_capacity, weights, values, memo)

    # option 2: take current item, if it fits
    if weights[index] <= remaining_capacity:
        taken = values[index] + solve(
            index + 1, remaining_capacity - weights[index], weights, values, memo
        )
        best = max(best, taken)

    memo[key] = best
    return best


def knapsack(weights, values, capacity):
    memo = {}
    return solve(0, capacity, weights, values, memo)


if __name__ == "__main__":
    w = [2, 2, 4, 3]
    v = [5, 3, 6, 4]
    c = 6
    print(knapsack(w, v, c))

File: server/datasets/java_source-code/Multiple-C2/organic_3.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #3
 * Top-down recursive solution with a manual memoization map.
 */

import java.util.HashMap;
import java.util.Map;

class Solution {
    private static int solve(int index, int remainingCapacity, int[] weights, int[] values, Map<String, Integer> memo) {
        if (index == weights.length || remainingCapacity == 0) {
            return 0;
        }

        String key = index + "," + remainingCapacity;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }

        int best = solve(index + 1, remainingCapacity, weights, values, memo);

        if (weights[index] <= remainingCapacity) {
            int taken = values[index] + solve(index + 1, remainingCapacity - weights[index], weights, values, memo);
            best = Math.max(best, taken);
        }

        memo.put(key, best);
        return best;
    }

    public static int knapsack(int[] weights, int[] values, int capacity) {
        Map<String, Integer> memo = new HashMap<>();
        return solve(0, capacity, weights, values, memo);
    }

    public static void main(String[] args) {
        int[] w = {2, 2, 4, 3};
        int[] v = {5, 3, 6, 4};
        int c = 6;
        System.out.println(knapsack(w, v, c));
    }
}
